gflownet/utils/batch.py

```diff
@@ -228,226 +333,870 @@ def states2policy(
+        elif isinstance(self.states, list):
+            states_term = [self.states[idx] for idx in indices if self.done[idx]]
+            if self.conditional and (policy is True or proxy is True):
+                done = np.array(self.done, dtype=bool)[indices]
+                traj_indices = np.array(self.traj_indices)[indices][done]
+                assert len(traj_indices) == len(np.unique(traj_indices))
+        else:
+            raise NotImplementedError("self.states can only be list or torch.tensor")
+        if policy is True:
+            return self.states2policy(states_term, traj_indices)
+        elif proxy is True:
+            return self.states2proxy(states_term, traj_indices)
+        else:
+            return states_term
+
+    def get_terminating_rewards(
+        self,
+        sort_by: str = "insertion",
+        force_recompute: Optional[bool] = False,
+    ) -> TensorType["n_trajectories"]:
+        """
+        Returns the reward of the terminating states in the batch, that is all states
+        with done = True. The returned rewards may be sorted by order of insertion
+        (sort_by = "insert[ion]", default) or by trajectory index (sort_by =
+        "traj[ectory]".
+
+        Args
+        ----
+        sort_by : str
+            Indicates how to sort the output:
+                - insert[ion]: sort by order of insertion (rewards of trajectories that
+                  reached the terminating state first come first)
+                - traj[ectory]: sort by trajectory index (the order in the ordered
+                  dict self.trajectories)
+
+        force_recompute : bool
+            If True, the rewards are recomputed even if they are available.
+        """
+        if sort_by == "insert" or sort_by == "insertion":
+            indices = np.arange(len(self))
+        elif sort_by == "traj" or sort_by == "trajectory":
+            indices = np.argsort(self.traj_indices)
+        else:
+            raise ValueError("sort_by must be either insert[ion] or traj[ectory]")
+        if self.rewards_available is False or force_recompute is True:
+            self._compute_rewards()
+        done = self.get_done()[indices]
+        return self.rewards[indices][done]
+
+    def get_actions_trajectories(self) -> List[List[Tuple]]:
+        actions_trajectories = []
+        for batch_indices in self.trajectories.values():
+            actions_trajectories.append([self.actions[idx] for idx in batch_indices])
+        return actions_trajectories
+
+    def get_states_of_trajectory(
+        self,
+        traj_idx: int,
+        states: Optional[
+            Union[TensorType["n_states", "..."], npt.NDArray[np.float32], List]
+        ] = None,
+        traj_indices: Optional[Union[List, TensorType["n_states"]]] = None,
+    ) -> Union[
+        TensorType["n_states", "state_proxy_dims"], npt.NDArray[np.float32], List
+    ]:
+        """
+        TODO: docstring
```

Comment:
Blasphemy!

Reply:
oops haha I wrote so much docs here to be fair!!!

Reply:
[There you go](https://github.com/alexhernandezgarcia/gflownet/pull/150/commits/ac8a8921e8b46d6bde7151634c40f42806a454a2)!